File: 1371-minimum-remove-to-make-valid-parentheses/solution.py
class Solution(object):
    def minRemoveToMakeValid(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = deque()
        removeIndices = set()
        
        for i in range(len(s)):
            if s[i] == ')':
                if len(stack) > 0:
                    stack.pop()
                else:
                    removeIndices.add(i)
                    
            elif s[i] == '(':
                stack.append((s[i], i))
        
        
        while len(stack):
            c, index = stack.pop()
            removeIndices.add(index)
        
        result = []
        for i in range(len(s)):
            if i not in removeIndices:
                result.append(s[i])
        return "".join(result)
                
            

File: 1371-minimum-remove-to-make-valid-parentheses/solution.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int balance = 0;
        stack<int> index_parenthesis;
        unordered_set<int> remove_idx;
        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == '(') {
                index_parenthesis.push(i);
            } else if (s[i] == ')') {
                if (index_parenthesis.empty()) {
                    remove_idx.insert(i);
                } else {
                    index_parenthesis.pop();
                }

            }
        }

        while (!index_parenthesis.empty()) {
            remove_idx.insert(index_parenthesis.top());
            index_parenthesis.pop();
        }

        string ans = "";
        for (int i = 0; i < s.size(); ++i) {
            if (remove_idx.find(i) == remove_idx.end()) {
                ans += s[i];
            }
        }

        return ans;
    }
};
